File: services/event-service/internal/adapter/messaging/kafka/consumer.go
package kafka

import (
	"context"
	"errors"
	"fmt"
	"strconv"
	"time"

	segkafka "github.com/segmentio/kafka-go"

	"github.com/buixuankhai1204/ticket-microservice-golang/services/event-service/internal/domain"
	"github.com/buixuankhai1204/ticket-microservice-golang/services/event-service/internal/platform/logger"
)

type Recorder[E any] interface {
	Execute(ctx context.Context, ev E) (alreadyProcessed bool, err error)
}

type EventSpec[E any] struct {
	Group      string
	EventType  string
	Component  string
	SuccessMsg string
	Parse      func([]byte) (E, error)
	LogFields  func(E) []any
	Record     Recorder[E]
}

type Config struct {
	Brokers     []string
	Topic       string
	MaxAttempts int
}

type Consumer[E any] struct {
	reader      *segkafka.Reader
	dlq         *segkafka.Writer
	spec        EventSpec[E]
	log         logger.Logger
	maxAttempts int
}

func NewConsumer[E any](cfg Config, spec EventSpec[E], log logger.Logger) *Consumer[E] {
	maxAttempts := cfg.MaxAttempts
	if maxAttempts < 1 {
		maxAttempts = 5
	}
	return &Consumer[E]{
		reader: segkafka.NewReader(segkafka.ReaderConfig{
			Brokers:        cfg.Brokers,
			GroupID:        spec.Group,
			Topic:          cfg.Topic,
			MinBytes:       1,
			MaxBytes:       10 * 1024 * 1024,
			MaxWait:        500 * time.Millisecond,
			CommitInterval: 0,
		}),
		dlq: &segkafka.Writer{
			Addr:         segkafka.TCP(cfg.Brokers...),
			Topic:        cfg.Topic + ".dlq",
			Balancer:     &segkafka.Hash{},
			RequiredAcks: segkafka.RequireAll,
		},
		spec:        spec,
		log:         log.With("component", spec.Component, "topic", cfg.Topic),
		maxAttempts: maxAttempts,
	}
}

func (c *Consumer[E]) Run(ctx context.Context) error {
	c.log.Info("consumer started", "group", c.spec.Group, "max_attempts", c.maxAttempts)

	for {
		m, err := c.reader.FetchMessage(ctx)
		if err != nil {
			if ctx.Err() != nil {
				c.log.Info("consumer stopping")
				return nil
			}
			c.log.Error("fetch failed, retrying", "err", err.Error())
			if sleep(ctx, time.Second) != nil {
				return nil
			}
			continue
		}

		if err := c.handle(ctx, m); err != nil {
			if ctx.Err() != nil {
				return nil
			}
			c.log.Error("message not processed, will be redelivered", "err", err.Error(), "offset", m.Offset)
			if sleep(ctx, time.Second) != nil {
				return nil
			}
			continue
		}

		if err := c.reader.CommitMessages(ctx, m); err != nil {
			if ctx.Err() != nil {
				return nil
			}
			c.log.Error("commit failed; message may be redelivered", "err", err.Error(), "offset", m.Offset)
		}
	}
}

func (c *Consumer[E]) Close() error {
	return errors.Join(c.reader.Close(), c.dlq.Close())
}

func (c *Consumer[E]) handle(ctx context.Context, m segkafka.Message) error {
	if t := headerValue(m, "event_type"); t != "" && t != c.spec.EventType {
		c.log.Info("event_type not handled by this consumer, skipping", "event_type", t, "offset", m.Offset)
		return nil
	}

	ev, parseErr := c.spec.Parse(m.Value)
	if parseErr != nil {
		c.log.Error("undeserializable message -> dlq", "err", parseErr.Error(), "offset", m.Offset)
		return c.toDLQ(ctx, m, "parse: "+parseErr.Error())
	}

	log := c.log.With(append(c.spec.LogFields(ev), "offset", m.Offset)...)
	backoff := 250 * time.Millisecond

	for attempt := 1; ; attempt++ {
		already, err := c.spec.Record.Execute(ctx, ev)
		switch {
		case err == nil:
			if already {
				log.Info("duplicate event skipped")
			} else {
				log.Info(c.spec.SuccessMsg)
			}
			return nil

		case ctx.Err() != nil:
			return ctx.Err()

		case isRetryable(err):
			if attempt >= c.maxAttempts {
				log.Error("max retries exhausted -> dlq", "attempts", attempt, "err", err.Error())
				return c.toDLQ(ctx, m, fmt.Sprintf("max-retries after %d attempts: %v", attempt, err))
			}
			log.Info("retryable error, backing off", "attempt", attempt, "backoff_ms", backoff.Milliseconds(), "err", err.Error())
			if sleep(ctx, backoff) != nil {
				return ctx.Err()
			}
			if backoff < 30*time.Second {
				backoff *= 2
			}

		default:
			log.Error("permanent error -> dlq", "err", err.Error())
			return c.toDLQ(ctx, m, "permanent: "+err.Error())
		}
	}
}

func (c *Consumer[E]) toDLQ(ctx context.Context, m segkafka.Message, reason string) error {
	dead := segkafka.Message{
		Key:   m.Key,
		Value: m.Value,
		Headers: []segkafka.Header{
			{Key: "x-dlq-reason", Value: []byte(reason)},
			{Key: "x-dlq-source-topic", Value: []byte(m.Topic)},
			{Key: "x-dlq-source-partition", Value: []byte(strconv.Itoa(m.Partition))},
			{Key: "x-dlq-source-offset", Value: []byte(strconv.FormatInt(m.Offset, 10))},
			{Key: "x-dlq-at", Value: []byte(time.Now().UTC().Format(time.RFC3339))},
		},
	}
	if err := c.dlq.WriteMessages(ctx, dead); err != nil {
		return fmt.Errorf("write to dlq: %w", err)
	}
	c.log.Error("message dead-lettered", "reason", reason, "offset", m.Offset)
	return nil
}

func isRetryable(err error) bool {
	var repoErr *domain.RepositoryError
	return errors.As(err, &repoErr)
}

func headerValue(m segkafka.Message, key string) string {
	for _, h := range m.Headers {
		if h.Key == key {
			return string(h.Value)
		}
	}
	return ""
}

func sleep(ctx context.Context, d time.Duration) error {
	t := time.NewTimer(d)
	defer t.Stop()
	select {
	case <-ctx.Done():
		return ctx.Err()
	case <-t.C:
		return nil
	}
}

File: services/booking-service/src/adapter/messaging/kafka/consumer.rs
use std::sync::Arc;
use std::time::Duration;

use async_trait::async_trait;
use rdkafka::config::ClientConfig;
use rdkafka::consumer::{CommitMode, Consumer, StreamConsumer};
use rdkafka::message::{BorrowedMessage, Header, Headers, Message, OwnedHeaders};
use rdkafka::producer::{FutureProducer, FutureRecord};
use serde::de::DeserializeOwned;
use tokio_util::sync::CancellationToken;

use crate::domain::{BookingError, SeatReserved};
use crate::usecase::ConfirmBookingUseCase;

pub enum HandlerError {
    Transient(String),
    Permanent(String),
}

#[async_trait]
pub trait SagaHandler: Send + Sync + 'static {
    type Event: DeserializeOwned + Send + Sync;

    fn group_id(&self) -> &str;
    fn event_type(&self) -> &str;
    async fn handle(&self, ev: &Self::Event) -> Result<bool, HandlerError>;
}

pub struct SagaConsumer<H: SagaHandler> {
    consumer: StreamConsumer,
    dlq: FutureProducer,
    dlq_topic: String,
    handler: H,
    max_attempts: u32,
}

impl<H: SagaHandler> SagaConsumer<H> {
    pub fn new(
        brokers: &str,
        topic: &str,
        max_attempts: u32,
        handler: H,
    ) -> Result<Self, rdkafka::error::KafkaError> {
        let consumer: StreamConsumer = ClientConfig::new()
            .set("bootstrap.servers", brokers)
            .set("group.id", handler.group_id())
            .set("enable.auto.commit", "false")
            .set("auto.offset.reset", "earliest")
            .create()?;
        consumer.subscribe(&[topic])?;

        let dlq: FutureProducer = ClientConfig::new()
            .set("bootstrap.servers", brokers)
            .create()?;

        Ok(Self {
            consumer,
            dlq,
            dlq_topic: format!("{topic}.dlq"),
            handler,
            max_attempts: max_attempts.max(1),
        })
    }

    pub async fn run(self, shutdown: CancellationToken) {
        tracing::info!(
            group = self.handler.group_id(),
            max_attempts = self.max_attempts,
            "kafka consumer started"
        );

        loop {
            let msg = tokio::select! {
                _ = shutdown.cancelled() => {
                    tracing::info!(group = self.handler.group_id(), "kafka consumer stopping");
                    return;
                }
                res = self.consumer.recv() => match res {
                    Ok(m) => m,
                    Err(e) => {
                        tracing::error!(err = %e, "kafka recv failed, retrying");
                        tokio::time::sleep(Duration::from_secs(1)).await;
                        continue;
                    }
                },
            };

            match self.process(&msg).await {
                Ok(()) => {
                    if let Err(e) = self.consumer.commit_message(&msg, CommitMode::Sync) {
                        tracing::error!(err = %e, "offset commit failed; message may be redelivered");
                    }
                }
                Err(e) => {
                    tracing::error!(err = %e, "message not processed; will be redelivered");
                    tokio::time::sleep(Duration::from_secs(1)).await;
                }
            }
        }
    }

    async fn process(&self, msg: &BorrowedMessage<'_>) -> Result<(), String> {
        if let Some(event_type) = header_str(msg, "event_type") {
            if event_type != self.handler.event_type() {
                tracing::info!(
                    event_type,
                    "event_type not handled by this consumer, skipping"
                );
                return Ok(());
            }
        }

        let payload = msg.payload().ok_or_else(|| "empty payload".to_string())?;
        let ev: H::Event = match serde_json::from_slice(payload) {
            Ok(ev) => ev,
            Err(e) => {
                tracing::error!(err = %e, "undeserializable message -> dlq");
                return self.to_dlq(msg, &format!("parse: {e}")).await;
            }
        };

        let mut backoff = Duration::from_millis(250);
        for attempt in 1..=self.max_attempts {
            match self.handler.handle(&ev).await {
                Ok(already) => {
                    if already {
                        tracing::info!("duplicate event skipped");
                    } else {
                        tracing::info!("saga event processed");
                    }
                    return Ok(());
                }
                Err(HandlerError::Transient(m)) => {
                    if attempt >= self.max_attempts {
                        tracing::error!(attempts = attempt, err = %m, "max retries exhausted -> dlq");
                        return self
                            .to_dlq(msg, &format!("max-retries after {attempt} attempts: {m}"))
                            .await;
                    }
                    tracing::warn!(attempt, backoff_ms = backoff.as_millis() as u64, err = %m, "transient error, backing off");
                    tokio::time::sleep(backoff).await;
                    backoff = (backoff * 2).min(Duration::from_secs(30));
                }
                Err(HandlerError::Permanent(m)) => {
                    tracing::error!(err = %m, "permanent error -> dlq");
                    return self.to_dlq(msg, &format!("permanent: {m}")).await;
                }
            }
        }

        Ok(())
    }

    async fn to_dlq(&self, msg: &BorrowedMessage<'_>, reason: &str) -> Result<(), String> {
        let partition = msg.partition().to_string();
        let offset = msg.offset().to_string();
        let headers = OwnedHeaders::new()
            .insert(Header {
                key: "x-dlq-reason",
                value: Some(reason),
            })
            .insert(Header {
                key: "x-dlq-source-topic",
                value: Some(msg.topic()),
            })
            .insert(Header {
                key: "x-dlq-source-partition",
                value: Some(partition.as_str()),
            })
            .insert(Header {
                key: "x-dlq-source-offset",
                value: Some(offset.as_str()),
            });

        let key: &[u8] = msg.key().unwrap_or(&[]);
        let payload: &[u8] = msg.payload().unwrap_or(&[]);
        let record = FutureRecord::to(&self.dlq_topic)
            .key(key)
            .payload(payload)
            .headers(headers);

        self.dlq
            .send(record, Duration::from_secs(5))
            .await
            .map_err(|(e, _)| format!("dlq send: {e}"))?;

        tracing::error!(reason, topic = %self.dlq_topic, "message dead-lettered");
        Ok(())
    }
}

fn header_str<'a>(msg: &'a BorrowedMessage<'a>, key: &str) -> Option<&'a str> {
    msg.headers()?
        .iter()
        .find(|h| h.key == key)?
        .value
        .and_then(|v| std::str::from_utf8(v).ok())
}

pub struct ConfirmBookingHandler {
    pub use_case: Arc<ConfirmBookingUseCase>,
}

#[async_trait]
impl SagaHandler for ConfirmBookingHandler {
    type Event = SeatReserved;

    fn group_id(&self) -> &str {
        "booking-service-SeatReserved"
    }

    fn event_type(&self) -> &str {
        "SeatReserved"
    }

    async fn handle(&self, ev: &SeatReserved) -> Result<bool, HandlerError> {
        self.use_case.execute(ev).await.map_err(|e| match e {
            BookingError::Repository(m) => HandlerError::Transient(m),
            other => HandlerError::Permanent(other.to_string()),
        })
    }
}
